rust/projects/networking/project.md

```diff
@@ -0,0 +1,288 @@
+
+
+## Part 3: Client-server networking setup
+
+Next we're going to set up the networking. For this project you are going to be
+using the basic TCP/IP networking APIs in `std::net`: `TcpListener` and
+`TcpStream`. Before thinking about the protocol, modify `kvs-server` to listen
+for and accept connections, and `kvs-client` to initiate connections.
+
+For this project, the server is synchronous and single-threaded. That means that
+you will listen on a socket, then accept connections, and execute and respond to
+commands one at a time. In the future we will re-visit this decision multiple
+times on our journey toward an asynchronous, multi-threaded, and
+high-performance database.
+
+Thank about your manual testing workflow. Now that there are two executables to
+deal with, you'll need a way to run them both at the same time. If you are like
+many, you will use two terminals, running `cargo run --bin --tikv-server` in
+one, where it runs until you press CTRL-D`, and `cargo run --bin --tikv-client`
+in the other.
+
+This is a good opportunity to use the logging macros for debugging. Go ahead and
+log information about every accepted connection.
+
+
+## Part 4: Implementing commands across the network
+
+Now it's time to implement the key/value store over the network, remotely
+executing commands that until now have been implemented within a single process.
+As with the file I/O you worked on in the last project to create the log, you
+will be serializing and streaming commands with the `Read` and `Write` traits.
+
+You are going to design a network protocol.
+
+TODO: Can't think of much to say about this
+
+
+## Part 5: More error handling
+
+- handle error responses by converting errors to a serializable format
+- add context to errors
+- replace `fn main() -> Result` with custom error reporting
+
+## Part 6: Pluggable storage engines
+
+Your database has a storage engine, `KvStore`, implemented by you.
+Now you are going to add a second storage engine.
+
+There are multiple reasons to do so:
+
+- Different workloads require difference performance characteristics. Some
+  storage engines may work better than other based on the workload.
+
+- It creates a familiar framework for comparing different backends.
+
+- It gives us an excuse to create and work with traits.
+
+- It gives us an excuse to write some comparative benchmarks!
+
+So you are going to _extract_ a new trait, `KvsEngine`, from the `KvStore`
+interface. This is a classic _refactoring_, where existing code is transformed
+into a new form incrementally. When refactoring you will generally want to break
+the work up into the smallest changes that will continue to build and work.
+
+Here is the API you need to end up with:
+
+- `trait KvsEngine` has `get` and `set` methods with the same signatures
+  as `KvStore`.
+
+- `KvStore` implements `KvsEngine`, and no longer has `get` and `set` methods of
+  its own.
+
+- There is a free function, `pub fn create_engine(name: &str) -> impl
```

Comment:
Inside this function, I think we have to use a `Box<dyn KvsEngine>` to make returning different types in one function possible. The dynamic dispatch is introduced here, and will happen every time a get/set command is called. However, the engine type is determined at the start of the server and will not change. I have not benchmarked it yet, and intution tells me that does not make much difference, but it still makes me feel uncomfortable. 

I want to avoid the unnecessary dynamic dispatch here and not to give instructions about this function. What do you think?

Reply:
Ah you are right. I also prefer not to do dynamic dispatch. I did though want to have an excuse for people to use impl trait. Please formulate this however you think is best and update the project text.